[FEATURE factory-for] Implement factoryFor

chadhietala opened the pull request:
Implements `lookup.factoryFor` to avoid double extending all objects.

See [RFC](https://github.com/emberjs/rfcs/blob/master/text/0150-factory-for.md)

chadhietala committed:
[FEATURE factory-for] Implement factoryFor

rwjblue commented:
/cc @mixonic

rwjblue commented:
Haven't had a chance to review in detail, but I did notice that this still needs feature flagging.

chadhietala commented:
@stefanpenner I think this good for first round of review. Currently no feature flagging has been done because these are effectively internal changes with no user-land implications.

rwjblue commented:
> Currently no feature flagging has been done because these are effectively internal changes with no user-land implications.

O_o, isn't the whole point of the RFC to expose this as public API?

mixonic commented:
@chadhietala 

> inject the deprecated container in this new public API

I'm not quite sure what this means, the container is not deprecated. Will review here.

homu commented:
:umbrella: The latest upstream changes (presumably #14355) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #14428) made this pull request unmergeable. Please resolve the merge conflicts.

mixonic commented on packages/container/lib/container.js:
```diff
@@ -127,9 +135,86 @@ Container.prototype = {
+
+    let manager = {
+      class: factory,
+      create(options = {}) {
```
I expected the implementation here to be

```js
let factory = this._lookupFactory(name);
return factory.create(options);
```

for the initial pass of `factoryFor` `create` per https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#detailed-design. Additionally https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#development-mode-proxy. In practice literally implementing as `_lookupFactory(name).create(props)` might not work, but it seems like it should and would preserve the double-extend behavior during instance creation. We should decide if this is really needed, will raise it in some conversations.

chadhietala replied:
Not sure if I follow. Can we not just transition all the internal stuff over to `factoryFor`?

mixonic replied:
The intent was to encourage libs/apps to use the new API and drop their current reliance on the double-extended class, then change over internals and any other small breaking changes that will happen after most apps have refactored toward `factoryFor` and stopped expecting a double-extend.

mixonic commented on packages/container/lib/container.js:
```diff
@@ -191,15 +280,47 @@ function lookup(container, fullName, options = {}) {
-  if (isSingleton(container, fullName) && options.singleton !== false) {
-    container.cache[fullName] = value;
+function isFactoryClass(container, fullname, { instantiate, singleton }) {
+  return (!isSingleton(container, fullname) || singleton === false) && (!shouldInstantiate(container, fullname) && instantiate === false);
```
putting the faster checks first seems ideal:

```js
function isFactoryClass(container, fullname, { instantiate, singleton }) {
  return (
    (singleton === false || !isSingleton(container, fullname)) &&
    (instantiate === false || !shouldInstantiate(container, fullname))
  );
}
```

thus avoiding the function calls in many cases.

mixonic commented on packages/container/lib/container.js:
```diff
@@ -127,9 +135,86 @@ Container.prototype = {
+      deprecate('Using "_lookupFactory" is deprecated. Please use container.factoryFor instead.', false, { id: 'container-lookupFactory', until: '2.12.0', url: 'TODO' });
+    }
+
     return factoryFor(this, this.registry.normalize(fullName), options);
```
can this function be renamed now, since it is not the `factoryFor` one would normally be looking for here.

mixonic commented on packages/container/lib/container.js:
```diff
@@ -191,15 +280,47 @@ function lookup(container, fullName, options = {}) {
   }
 
-  let value = instantiate(container, fullName);
+  return instantiateFactory(container, fullName, options);
```
So this means`container.lookup` is now based on the the non-double-extended factory? This seems like the behavior we didn't intend to change until `_lookupFactory` is removed.

stefanpenner replied:
yes, we should only have 1 behavior in the system at any point in time, once we cut over it should be all no double extend.

mixonic commented on packages/container/lib/container.js:
```diff
@@ -127,9 +135,86 @@ Container.prototype = {
     assert('fullName must be a proper full name', this.registry.validateFullName(fullName));
+
+    if (isFeatureEnabled('container-factoryFor')) {
+      deprecate('Using "_lookupFactory" is deprecated. Please use container.factoryFor instead.', false, { id: 'container-lookupFactory', until: '2.12.0', url: 'TODO' });
```
Must be `until: '2.13'`. 2.12 is an LTS release and must contain this deprecation.

mixonic commented on packages/container/tests/container_test.js:
```diff
@@ -15,6 +15,19 @@ QUnit.module('Container', {
+function lookupFactory(name, container, options) {
+  let factory;
+  if (isFeatureEnabled('container-factoryFor')) {
+    expectDeprecation(() => {
```
There is also an `ignoreDeprecation` which might be appropriate/useful here. Example:

https://github.com/emberjs/ember.js/blob/aeeddf7c00b42a340494465252481b5414d87009/packages/ember-runtime/tests/controllers/controller_test.js#L118

mixonic commented on packages/container/tests/container_test.js:
```diff
@@ -561,7 +581,8 @@ QUnit.test('A deprecated `container` property is appended to every object instan
 
-QUnit.test('A deprecated `container` property is appended to every object instantiated from a non-extendable factory, and a fake container is available during instantiation.', function() {
+// This is testing that container was passed as an option
+QUnit.skip('A deprecated `container` property is appended to every object instantiated from a non-extendable factory, and a fake container is available during instantiation.', function() {
```
This test shows why we need to keep the "double extend" the default path for `lookup` until at least 2.13.

mixonic commented on packages/container/lib/container.js:
```diff
@@ -127,9 +135,86 @@ Container.prototype = {
+            throw new Error(`You attempted to set "${prop}" on a factory manager created by container#factoryFor. A factory manager is a read-only construct.`);
+          }
+        };
+        manager = new Proxy(manager, validator);
```
Would like to have tests included that:

* getting a property that is not `class` or `create` throws an error
* setting a property throws an error
* calling `instanceof` with the return value from `factoryFor` throws an error

Per: https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#development-mode-proxy

mixonic reviewed (changes requested):
Thanks so much for getting this going @chadhietala. The big item is that `lookup` should still be using double extend. If you want to add a second feature flag (ideally in a followup) that experimentally switches `lookup` to the non-double-extended factory, that would definitely be welcome as a distinct feature flag.

mixonic commented on packages/ember-extension-support/lib/data_adapter.js:
```diff
@@ -163,7 +163,7 @@ export default EmberObject.extend({
   _nameToClass(type) {
     if (typeof type === 'string') {
-      type = getOwner(this)._lookupFactory(`model:${type}`);
+      type = getOwner(this).factoryFor(`model:${type}`).class;
```
It seems there is some use of the types in Ember-Data. For example [here attributes are read off the class](https://github.com/emberjs/data/blob/c9d8212c857ca78218ad98d11621819b38dba98f/addon/-private/system/debug/debug-adapter.js#L32). Removing the double-extend means the `attributes` CP would be called and memoized on the base class, and thus shared between test runs etc.

This is definitely a change, described [in the design section](https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#design-of-applicationinstancefactoryfor).

Some other examples of libraries this would impact are listed [in the drawbacks section](https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#drawbacks).

homu commented:
:umbrella: The latest upstream changes (presumably #14487) made this pull request unmergeable. Please resolve the merge conflicts.

chadhietala commented:
I made some changes to this, but likely needs further guidance.

homu commented:
:umbrella: The latest upstream changes (presumably #14545) made this pull request unmergeable. Please resolve the merge conflicts.

stefanpenner commented:
reviewing now

stefanpenner commented on packages/container/lib/container.js:
```diff
@@ -127,7 +136,96 @@ Container.prototype = {
+            validationCache[fullName] = true;
+          });
+
+          this.class[NAME_KEY] = container.registry.makeToString(factory, fullName);
```
I don't believe we want to mutate the `class`, note: it is perfectly legal to register a class multiple times with different injection rules/similar for each. When this occurs, nothing should collide and currently NAME_KEY would collide. I believe NAME_KEY can be provided merely as a member of `props`. I believe in most cases this should work IFF

https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/system/core_object.js#L546

changed to:

```js
var ret = '<' + (this[NAME_KEY] || this.constructor.toString()) + ':' + guidFor(this) + extension + '>';
```

This would mean, `class.toString()` wouldn't be terribly useful, but I don't believe it can be accomplished

stefanpenner commented on packages/container/lib/container.js:
```diff
@@ -127,7 +136,96 @@ Container.prototype = {
+          }
+
+          if (this.class.prototype) {
+            injectDeprecatedContainer(this.class.prototype, container);
```
I don't believe we want to mutate the class

stefanpenner commented on packages/container/lib/container.js:
```diff
@@ -350,7 +474,7 @@ function instantiate(container, fullName) {
 
     if (typeof factory.extend === 'function') {
```
does this refer to the `factory.class` or our new factory object, if it is our new factory object this may no longer make sense?

mixonic replied:
This refers to `extend` on the registered class, not `extend` on the factory object. The idea being `extend` is supported, injections have been made.

`instantiate` is only called by `lookup` though, and in that case we want the double extend to continue for now so this test would hold true.

stefanpenner commented on packages/container/lib/container.js:
```diff
@@ -374,6 +498,19 @@ function instantiate(container, fullName) {
+  let factoryInjections = buildInjections(container,
+                                          registry.getFactoryTypeInjections(type),
+                                          registry.getFactoryInjections(fullName));
+  factoryInjections._debugContainerKey = fullName;
```
@mixonic  / @rwjblue should this become a symbol or something in the future?

mixonic replied:
In the future perhaps. I don't think it should be changed in this effort.

stefanpenner committed:
[FEATURE factory-for] Make helpers work with factoryFor

stefanpenner commented on packages/ember-routing/lib/system/generate_controller.js:
```diff
@@ -17,7 +17,7 @@ import {
 
 export function generateControllerFactory(owner, controllerName, context) {
-  let Factory = owner._lookupFactory('controller:basic').extend({
+  let Factory = owner.factoryFor('controller:basic').class.extend({
```
if previously, injections where set on `controller:basic` they would now be lost (as we bypass DI if we do `.class`

mixonic replied:
@chadhietala and I went over this. The change as it stands would indeed squash DI in the case where there was an injection on `controller:basic` and you expected it to show up on `controller:generated-controller`. Of course that whole concept, that your DI onto `controller:basic` will show up on `controller:generated-controller`, is pretty WAT. But we should preserve it if possible.

stefanpenner commented on packages/ember-routing/lib/system/route.js:
```diff
@@ -1631,16 +1632,15 @@ let Route = EmberObject.extend(ActionHandler, Evented, {
+        assert(`${StringUtils.classify(name)} has no method \`find\`.`, typeof modelClass.class.find === 'function');
 
-        return modelClass.find(value);
+        return modelClass.class.find(value);
```
In a future iteration, we should consider deprecating this feature. it is... dubious at best (and was my fault...)

stefanpenner commented on packages/ember-routing/lib/system/router.js:
```diff
@@ -585,7 +585,7 @@ const EmberRouter = EmberObject.extend(Evented, {
 
       if (!handler) {
-        let DefaultRoute = routeOwner._lookupFactory('route:basic');
+        let DefaultRoute = routeOwner.factoryFor('route:basic').class;
```
similar to the above `controller:basic` example, if injections existed on `route:basic` at this point they would be lost, which I believe is a breaking change.

chadhietala replied:
Same question as above.

mixonic replied:
If we care about injections on `route:basic` being available on all generated routes, we can take the current implementation:

```js
let DefaultRoute = routeOwner._lookupFactory('route:basic');
routeOwner.register(fullRouteName, DefaultRoute.extend());
handler = routeOwner.lookup(fullRouteName);
// Later on may lookup this name again
```

And hack a version that should do the double-DI (DI for `route:basic` and `route:generated-thing` with the current implementation:

```js
let factory = routeOwner.factoryFor('route:basic');
let DefaultRouteProxy = {
  extend(...diArgs) {
    return {
      create(...createArgs) {
         return factory.create(assign({}, ...diArg, ...createArgs));
      }
    };
  },
  create(...createArgs) {
    /* Not sure this is ever going to be used in the current container double-extend implementation */
    return factory.create(assign({}, ...createArgs));
  }
}
```

However in 99.99999% of cases I don't expect someone is doing an injection on `route:basic` (or `controller:basic`), instead they are probably doing injections on the type. To keep these cases fast, we can check the injections list and if possible simple re-register the class:

```js
let DefaultRoute = routeOwner.factoryFor('route:basic').class;
routeOwner.register(fullRouteName, DefaultRoute);
handler = routeOwner.lookup(fullRouteName);
```

In this case DI on the type still works.

chadhietala replied:
So this doesn't seem to work for generated controllers due to the fact `.proto()` called right after we generate the controller. https://github.com/emberjs/ember.js/blob/192017b71070d5a32f24ec215ad75dd702d0c0bc/packages/ember-routing/lib/system/route.js#L193

chadhietala committed:
[FEATURE factory-for] Introduce internal FACTORY_FOR symbol method

homu commented:
:umbrella: The latest upstream changes (presumably #14582) made this pull request unmergeable. Please resolve the merge conflicts.

mixonic commented on packages/ember-routing/lib/system/generate_controller.js:
```diff
@@ -17,7 +17,7 @@ import {
 export function generateControllerFactory(owner, controllerName, context) {
-  let Factory = owner._lookupFactory('controller:basic').extend({
+  let Factory = owner.factoryFor('controller:basic').class.extend({
     isGenerated: true,
```
`isGenerated` can be killed off.

mixonic commented on packages/container/lib/container.js:
```diff
@@ -466,3 +575,56 @@ function buildFakeContainerFunction(container, containerProperty, ownerProperty)
+      validationCache[this.fullName] = true;
+    });
+
+    this.class[NAME_KEY] = this.container.registry.makeToString(this.class, this.fullName);
```
It does looks like this should be added to the injections [per master](https://github.com/emberjs/ember.js/blob/e601285bed0e2e108d41a68a61dc2098aac26dd7/packages/container/lib/container.js#L273) instead of set on the class.

mixonic commented:
Summary comment of our planning:

* [ ] There should be two feature flags. One adding `factoryFor` with the legacy double-extend implementation, one enabling no-more-double-extend mode.
* [ ] `isGenerated` is no longer needed as a flag on generated classes. (@rwjblue sanity check me?)
* [ ] In the `factoryFor` documentation we should note that by using `factoryFor` the consuming code takes responsibility for destroying the instance.
* [ ] It seems we can solve the injections-on-generated-singletons (routes, controllers) issue by continuing to use the register hack already in master, but we should also implement an above noted meta-hack to keep injections on `route:basic` and `controller:basic` showing up on all generated instances. I actually think this behavior is a) wrong and b) not relied upon and would prefer to just nuke it, but it warrants discussion.
* [ ] `NAME_KEY` needs fixin'
* [ ] `_nameToClass` in the data_adapter seems like it should return either the double extended class or the registered class depending on if the second no-more-double-extend feature flag is enabled.

rwjblue commented:
> isGenerated is no longer needed as a flag on generated classes. (@rwjblue sanity check me?)

Confirm.

chadhietala commented:
@mixonic so I think there is some problems with the last 2 items. The `_nameToClass` item above can't actually be tested sense flags are turned on in binary fashion. I can put the guard in there but we are basically flying blind on the `no-double-extend` path. The generated controllers and routes almost work however in the controller case we try to call `proto()` on the generated class so doing the proxying class hack doesn't really work.

homu commented:
:umbrella: The latest upstream changes (presumably #14633) made this pull request unmergeable. Please resolve the merge conflicts.

chadhietala commented on packages/container/lib/container.js:
```diff
@@ -28,6 +39,9 @@ export default function Container(registry, options) {
   this.isDestroyed = false;
 }
 
+Container.__FACTORY_FOR__ = FACTORY_FOR;
```
This seems bad but don't know how to make the Node tests pass without doing this.

mixonic commented on packages/container/lib/container.js:
```diff
@@ -127,7 +141,73 @@ Container.prototype = {
     assert('fullName must be a proper full name', this.registry.validateFullName(fullName));
-    return factoryFor(this, this.registry.normalize(fullName), options);
+
+    if (isFeatureEnabled('ember-no-double-extend')) {
```
I think this check should be `ember-factory-for`?

Additionally it seems like you would pass the boolean as an argument to `deprecate`, doesn't really matter though :-p

mixonic commented on packages/container/lib/container.js:
```diff
@@ -175,10 +255,31 @@ Container.prototype = {
 
+if (isFeatureEnabled('ember-factory-for')) {
+  /**
+   Given a fullName, return the corresponding factory. The consumer of factory is repsonsible
```
"of factory" -> "of the factory"

"repsonsible" -> "responsible"

mixonic commented on packages/container/lib/container.js:
```diff
@@ -175,10 +255,31 @@ Container.prototype = {
+if (isFeatureEnabled('ember-factory-for')) {
+  /**
+   Given a fullName, return the corresponding factory. The consumer of factory is repsonsible
+   for the destruction of the factory as there is no way to understand the objects lifecycle.
```
"for the destruction of the factory" -> "for the destruction of any factory instances, as there is no way for the container to ensure instances are destroyed when it itself is destroyed"

mixonic commented on packages/container/tests/container_test.js:
```diff
@@ -15,6 +15,19 @@ QUnit.module('Container', {
+function lookupFactory(name, container, options) {
+  let factory;
+  if (isFeatureEnabled('ember-no-double-extend')) {
+    expectDeprecation(() => {
```
`ignoreDeprecation` may be more appropriate for a test utility method.

mixonic commented on packages/container/tests/container_test.js:
```diff
@@ -465,15 +481,19 @@ QUnit.test('factory for non extendables resolves are cached', function() {
 QUnit.test('The `_onLookup` hook is called on factories when looked up the first time', function() {
-  expect(2);
+  if (isFeatureEnabled('ember-factory-for') && isFeatureEnabled('ember-no-double-extend')) {
+    expect(4);
```
yeah `ignoreDeprecation` would allow you to not have this change in expect count for an unrelated test.

mixonic commented on packages/ember-application/tests/system/dependency_injection/default_resolver_test.js:
```diff
@@ -64,9 +65,15 @@ QUnit.test('the default resolver looks up templates in Ember.TEMPLATES', functio
+    equal(locator.factoryFor('template:fooBar').class, fooBarTemplate, 'resolves template:foo_bar');
+    equal(locator.factoryFor('template:fooBar.baz').class, fooBarBazTemplate, 'resolves template:foo_bar.baz');
+  } else {
+    equal(locator.lookupFactory('template:foo'), fooTemplate, 'resolves template:foo');
```
It is concerning that this test coverage is lost when the feature flag is enabled?

mixonic commented on packages/ember-glimmer/lib/syntax/render.js:
```diff
@@ -186,8 +187,9 @@ class NonSingletonRenderManager extends AbstractRenderManager {
   create(environment, definition, args, dynamicScope) {
     let { name, env } = definition;
     let modelRef = args.positional.at(0);
+    let controllerFactory = env.owner[FACTORY_FOR](`controller:${name}`);
```
Doesn't using factoryFor here mean the double extended class is never used for `render`, regardless of the feature flag setting?

mixonic committed:
[FEATURE factory-for] Allow for incremental roll out factoryFor

The public factoryFor API should always return `class` as the
no-double-extend version. However internal APIs already refactored to
use FACTORY_FOR should use the double extended version if the
no-double-extend feature flag is false.

mixonic commented:
OMG this is good to go. To merge:

* One of us will rebase this down to a more meaningful commit history with `[FEATURE factory-for]` in the commits
* I would like to boot the build with various flags enabled/disabled in a real world app before merge. Just a quick sanity check.

Excited :-D

chadhietala commented:
Quick benchmarks from Travis-Web. This is consistent across runs and just looking at the homepage that is likely pretty static.

__Before: (Prod / Feature Flags Off)__
<img width="1005" alt="screen shot 2016-12-17 at 7 31 51 pm" src="https://cloud.githubusercontent.com/assets/183799/21291338/cd026c5a-c48f-11e6-8eeb-2eabf9d7a56e.png">

__After: (Prod / Feature Flags On)__
<img width="1003" alt="screen shot 2016-12-17 at 7 32 51 pm" src="https://cloud.githubusercontent.com/assets/183799/21291339/daa5e508-c48f-11e6-8016-35270c6e725a.png">

rwjblue commented:
@chadhietala - I assume the `After: (Prod / Feature Flags On)` screenshots above are with both `ember-factory-for` and `ember-no-double-extend` flags on, is that right?  Can you test with only `ember-factory-for` (I'm mostly looking to try to figure out if we will have temporary regressions in 2.12)?

rwjblue commented on packages/ember-application/lib/system/engine-instance.js:
```diff
@@ -194,9 +194,25 @@ const EngineInstance = EmberObject.extend(RegistryProxyMixin, ContainerProxyMixi
     this.inject('route', '_environment', '-environment:main');
+  },
+
+  [FACTORY_FOR](fullName, options) {
```
Lets make this part of the `ContainerProxyMixin`, this would simplify issues in `ember-test-helpers` land and actually keep the logic in the correct place (currently `.lookup` and `._lookupFactory` are done in `ContainerProxyMixin`).

rwjblue commented on packages/ember-application/lib/system/engine-instance.js:
```diff
@@ -194,9 +194,25 @@ const EngineInstance = EmberObject.extend(RegistryProxyMixin, ContainerProxyMixi
 
+if (isFeatureEnabled('ember-factory-for')) {
+  EngineInstance.reopen({
+    factoryFor(fullName, options) {
```
This can be moved to the mixin also...

rwjblue commented on packages/container/lib/container.js:
```diff
@@ -1,8 +1,19 @@
+import { assert, deprecate, runInDebug, isFeatureEnabled } from 'ember-metal';
 
 const CONTAINER_OVERRIDE = symbol('CONTAINER_OVERRIDE');
+const HAS_PROXY = typeof Proxy === 'function';
```
Can we move this into `ember-utils` and export as `HAS_NATIVE_PROXY`?  Very similar to what was done in https://github.com/emberjs/ember.js/pull/14649 for `packages/ember-utils/lib/weak-map-utils.js`.

rwjblue commented on tests/node/helpers/build-owner.js:
```diff
@@ -1,5 +1,14 @@
 module.exports = function buildOwner(Ember, resolver) {
-  var Owner = Ember.Object.extend(Ember._RegistryProxyMixin, Ember._ContainerProxyMixin);
+  var FACTORY_FOR = Ember.Container.__FACTORY_FOR__;
```
See comments in https://github.com/emberjs/ember.js/pull/14360#pullrequestreview-13465262 and https://github.com/emberjs/ember-test-helpers/pull/191, we should be able to remove this.

chadhietala committed:
[FEATURE factory-for] Add symboled methods to container proxy

mixonic committed:
[FEATURE factory-for] Only expose factoryFor w/ FF

mixonic commented:
I've done some benchmarks against todomvc. I would expect the impact on a larger app to be much more significant (as @chadhietala's profiling shows).

This PR may have a minor performance regression:

<img width="608" alt="screen shot 2016-12-18 at 2 19 48 pm" src="https://cloud.githubusercontent.com/assets/8752/21297032/1fad1250-c52d-11e6-9544-0d6cc0c82e6d.png">

The `ember-factory-for` feature flag does not seem to have a significant impact on performance (which is expected).

However the `ember-no-double-extend` feature flag makes a significant improvement.

<img width="610" alt="screen shot 2016-12-18 at 2 22 14 pm" src="https://cloud.githubusercontent.com/assets/8752/21297056/78ed13e2-c52d-11e6-89bc-55d697a9dcf4.png">

I think we can continue to explore issues around backwards compatibility, performance, and the migration path to `factoryFor` on master. Merging this.

@chadhietala Thanks much for your patience and push on this gnarly, complex, cross-cutting part of Ember's API. The payoff (especially in big apps) is going to be really notable.

stefanpenner commented:
👍

sandstrom commented:
Awesome! Nice to be able to move into a public API for this + great performance improvements! ⛵️
